### ISSUE TITLE
Possibility of writing "no-useless-ts-ignore-comments" rule 

### ISSUE THREAD

**Issue by mohsen1** [typescript-eslint/typescript-eslint#236]
I'm exploring if it's possible to write a rule that warns against `// @ts-ignore` comments that have become stale as result of changes in other places of even on the line that tsc is being ignored. My approach is to get all diagnostics and cross reference them with `// @ts-ignore` comments and see which comment does not actually ignoring an error. But if the `// @ts-ignore` comment is still there TypeScript will not report any diagnostics for the next line.

**Is there any way of getting diagnostics _if_ we remove the `// @ts-ignore` comment?**

One solution is to remove all `// @ts-ignore` comments from source files of existing program, construct a new program with those source files and get diagnostics. This seems very very expensive 

Any thoughts on this?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.3.0` |
| `TypeScript`                       | `3.0` |

**Comment by bradzacher** [comment]
it might be worth looking at https://github.com/mysticatea/eslint-plugin-eslint-comments
It does a similar thing except with eslint comments.

Glancing at that code makes me suspect this is something we can't handle purely within the plugin?

**Comment by platinumazure** [comment]
Just spitballing: It seems like it could be possible if TSC includes all errors/warnings, including the ignored warnings under a different object property. Then it would be easy to check those warnings' locations against the TS ignore comments and see if any of the ignore comments have no corresponding warnings. But not sure if that's the intended/ideal direction, or if TSC should be handling this.

**Comment by bradzacher** [comment]
The problem there is that we don't attach comments to nodes, and eslint deprecated the functionality to do that. https://github.com/typescript-eslint/typescript-eslint/issues/158#issuecomment-458737802

So whilst we know about the comments, we don't know what node they're attached to (unless tsc exposes that information). This means we don't know which node to check for errors (even if tsc did surface the ignored errors for us).

For that reason alone, I would say that this rule would be non-trivial to implement.

**Comment by platinumazure** [comment]
@bradzacher You wouldn't need to use comment attachment- you would just need to compare range or location information. That said, I agree that this would, in general, not be trivial to implement.

**Comment by mohsen1** [comment]
`//@ts-ignore` comments ignore _all_ type errors on next line. Something like [this is possible](https://www.typescriptlang.org/play/index.html#src=%2F%2F%20%40ts-ignore%0D%0Alet%20a%3A%20string%20%3D%201%3B%20let%20b%3A%20number%20%3D%20'str'):

```ts
// @ts-ignore
let a: string = 1; let b: number = 'str'
```

I agree with @platinumazure that TSC should provide information about ignored diagnostics so we can cross check them with the comments. Made an issue for that in TypeScript:
https://github.com/Microsoft/TypeScript/issues/29843

**Comment by JoshuaKGoldberg** [comment]
Relevant: https://github.com/microsoft/TypeScript/pull/36014 will add a `// @ts-expect-error` to TypeScript. We could make this rule instead just ban `// @ts-ignore` altogether.

**Comment by bradzacher** [comment]
It's already a rule (as of a week or so)! Just missed this issue.
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-ts-comment.md

**Comment by mohsen1** [comment]
This is different than what I suggest. I wanted to make sure we don’t have dead `ts-ignore` directives. Banning it altogether is not the same. 

This is useful when you have an issue that you can’t resolve and you use ts ignore to surpass it but then later via some mechanism like dependency update the type error is no longer there but the ts-ignore comment is there and it doesn’t allow ts to check the next line.

**Comment by bradzacher** [comment]
Relevant ts issue: https://github.com/microsoft/TypeScript/issues/29394

**Comment by mohsen1** [comment]
this can basically solve the problem 
https://github.com/sveltejs/svelte/issues/4518

**Comment by bradzacher** [comment]
This can be closed now, as microsoft/TypeScript#36014 has been merged, and will be released as part of the next TS version.

You'll have to migrate your codebases to use `ts-expect-error` when it's released, and then any comments that don't cause a TS error will themselves cause a TS error.
